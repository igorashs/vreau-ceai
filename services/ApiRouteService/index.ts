import ApiRouteService from './ApiRouteService';

export default ApiRouteService;
