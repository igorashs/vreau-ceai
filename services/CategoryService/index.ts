import CategoryModel from '@/models/Category';
import CategoryService from './CategoryService';

export default CategoryService(CategoryModel);
