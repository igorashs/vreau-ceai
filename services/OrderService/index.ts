import OrderModel from '@/models/Order';
import OrderService from './OrderService';

export default OrderService(OrderModel);
