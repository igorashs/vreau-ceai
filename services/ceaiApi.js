const URL = process.env.NEXT_PUBLIC_API_URL;

export const login = async (data) => {
  const res = await fetch(`${URL}/api/users/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const signup = async (data) => {
  const res = await fetch(`${URL}/api/users/signup`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const logout = async () => {
  const res = await fetch(`${URL}/api/users/logout`, {
    method: 'POST'
  });

  const resData = await res.json();

  return resData;
};

export const findUser = async (email) => {
  const res = await fetch(`${URL}/api/users?search=${email}`);

  const resData = await res.json();

  return resData;
};

export const updateUserManagerPermission = async (id, isManager) => {
  const res = await fetch(`${URL}/api/users/${id}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ isManager })
  });

  const resData = await res.json();

  return resData;
};

export const refreshUser = async () => {
  const res = await fetch(`${URL}/api/users/refresh`, {
    method: 'POST',
    credentials: 'same-origin'
  });

  const resData = await res.json();

  return resData;
};

export const createCategory = async (data) => {
  const res = await fetch(`${URL}/api/categories`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const getCategories = async () => {
  const res = await fetch(`${URL}/api/categories`);

  const resData = await res.json();

  return resData;
};

export const findCategory = async (name) => {
  const res = await fetch(`${URL}/api/categories?search=${name}`);

  const resData = await res.json();

  return resData;
};

export const updateCategory = async (id, data) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const deleteCategory = async (id) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin'
  });

  const resData = await res.json();

  return resData;
};
