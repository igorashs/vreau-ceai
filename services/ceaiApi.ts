import {
  Category,
  ApiResponse,
  Product,
  UserLogin,
  UserSignup,
  UserEmail,
  User,
  UserPermissions,
  OrderFields,
  UserOrders,
  ProductWithCategory,
  ProductErrorDetail,
  ProductName,
  ManagementOrders,
  OrderStatus,
  Order,
  OrderStatusErrorDetail,
  OrderNumber,
  OrderNumberErrorDetail,
  UserEmailErrorDetail,
  ProductNameErrorDetail,
  CategoryName,
  CategoryNameErrorDetail,
  SessionAuth,
} from 'types';

const URL = process.env.NEXT_PUBLIC_API_URL;

export const login = async <P extends UserLogin, R extends ApiResponse>(
  data: P,
): Promise<R> => {
  const res = await fetch(`${URL}/api/users/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const signup = async <P extends UserSignup, R extends ApiResponse>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/users/signup`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const logout = async <R extends ApiResponse>() => {
  const res = await fetch(`${URL}/api/users/logout`, {
    method: 'POST',
  });

  return res.json() as Promise<R>;
};

export const findUser = async <
  P extends UserEmail,
  R extends ApiResponse & { user: User; errors?: UserEmailErrorDetail[] }
>({
  email,
}: P) => {
  const res = await fetch(`${URL}/api/users?search=${email}`);

  return res.json() as Promise<R>;
};

export const updateUserManagerPermission = async <
  P extends UserPermissions,
  R extends ApiResponse & { user: User }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/users/${id}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const refreshUser = async <
  R extends ApiResponse & { session: SessionAuth }
>() => {
  const res = await fetch(`${URL}/api/users/refresh`, {
    method: 'POST',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const createCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/categories`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const getCategories = async <
  R extends ApiResponse & { categories: Category[] }
>() => {
  const res = await fetch(`${URL}/api/categories`);

  return res.json() as Promise<R>;
};

export const findCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>({
  name,
}: P) => {
  const res = await fetch(`${URL}/api/categories?search=${name}`);

  return res.json() as Promise<R>;
};

export const updateCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'PUT',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const deleteCategory = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const createProduct = async <
  R extends ApiResponse & { product: Product; errors?: ProductErrorDetail[] }
>(
  data: FormData,
) => {
  const res = await fetch(`${URL}/api/products/create`, {
    method: 'POST',
    credentials: 'same-origin',
    body: data,
  });

  return res.json() as Promise<R>;
};

export const updateProduct = async <
  R extends ApiResponse & { product: Product; errors?: ProductErrorDetail[] }
>(
  id: string,
  data: FormData,
) => {
  const res = await fetch(`${URL}/api/products/update/${id}`, {
    method: 'PUT',
    credentials: 'same-origin',
    body: data,
  });

  return res.json() as Promise<R>;
};

export const deleteProduct = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/products/delete/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const findProduct = async <
  R extends ApiResponse & {
    product: Product;
    errors?: ProductNameErrorDetail[];
  }
>({
  name,
}: ProductName) => {
  const res = await fetch(`${URL}/api/products?search=${name}`);

  return res.json() as Promise<R>;
};

export const getProducts = async <
  P extends string[] | null,
  R extends ApiResponse & { products: Product[]; count: number }
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};

export const getProductsByCategory = async <
  P extends string[] | null,
  R extends ApiResponse & { products: Product[]; count: number }
>(
  name: string,
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products?search=${name}&byCategory=true${
      filters ? `&filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};

export const createOrder = async <
  P extends OrderFields,
  R extends ApiResponse & { number: string }
>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/orders/create`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const findOrder = async <
  R extends ApiResponse & { order: Order; errors?: OrderNumberErrorDetail[] }
>({
  number,
}: OrderNumber) => {
  const res = await fetch(`${URL}/api/orders?search=${number}`, {
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const updateOrder = async <
  P extends OrderStatus,
  R extends ApiResponse & { order: Order; errors?: OrderStatusErrorDetail[] }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/orders/update/${id}`, {
    credentials: 'same-origin',
    method: 'PUT',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const deleteOrder = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/orders/delete/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const getOrders = async <
  P extends string[] | null,
  R extends ApiResponse & ManagementOrders
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/orders${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
    { credentials: 'same-origin' },
  );

  return res.json() as Promise<R>;
};

export const getUserOrders = async <
  P extends string[] | null,
  R extends ApiResponse & UserOrders
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/orders${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}&byUserId=true`,
    { credentials: 'same-origin' },
  );

  return res.json() as Promise<R>;
};

export const getRecommendedProducts = async <
  R extends ApiResponse & { products: ProductWithCategory[] }
>(
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products?recommended=true&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};
