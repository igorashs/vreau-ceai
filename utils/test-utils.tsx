import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
