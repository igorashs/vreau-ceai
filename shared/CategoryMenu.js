import styled from 'styled-components';
import { StyledLink } from './StyledLink';
import { DropDownList } from './DropDownList';
import { useState, useEffect } from 'react';
import { getCategories } from 'services/ceaiApi';

const Wrapper = styled.div`
  h5 {
    margin-bottom: calc(var(--baseline) / 4);
  }
`;

export function CategoryMenu() {
  const [dbCategories, setDbCategories] = useState();

  useEffect(async () => {
    const res = await getCategories();

    if (res.success) {
      setDbCategories(res.categories);
    }
  }, []);

  return (
    <Wrapper>
      <h5>Categorii</h5>
      {dbCategories && (
        <DropDownList label="Ceai">
          {dbCategories.map((c) => (
            <li key={c._id}>
              <StyledLink
                href={`/categories/${c.name}`}
                text={c.name}
                accent="dark"
              />
            </li>
          ))}
        </DropDownList>
      )}
    </Wrapper>
  );
}
