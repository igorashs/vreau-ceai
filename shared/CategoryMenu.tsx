import styled from 'styled-components';
import { useState, useEffect } from 'react';
import { getCategories } from 'services/ceaiApi';
import { Category } from 'types';
import StyledLink from './StyledLink';
import DropDownList from './DropDownList';

const Wrapper = styled.div`
  h5 {
    margin-bottom: calc(var(--baseline) / 4);
  }
`;

const CategoryMenu = () => {
  const [dbCategories, setDbCategories] = useState<Category[]>();

  useEffect(() => {
    const fetchData = async () => {
      const res = await getCategories();

      if (res.success) {
        setDbCategories(res.categories);
      }
    };

    fetchData();
  }, []);

  return (
    <Wrapper>
      <h5>Categorii</h5>
      {dbCategories && (
        <DropDownList label="Ceai">
          {dbCategories.map((c) => (
            <li key={c._id}>
              <StyledLink
                label={`Categoria ${c.name}`}
                href={`/categories/${c.name}`}
                text={c.name}
                accent="dark"
              />
            </li>
          ))}
        </DropDownList>
      )}
    </Wrapper>
  );
};

export default CategoryMenu;
