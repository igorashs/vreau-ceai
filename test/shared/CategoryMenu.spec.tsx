import { render, waitFor } from '@/utils/test-utils';
import CategoryMenu from '@/shared/CategoryMenu';
import { getCategories } from 'services/ceaiApi';

const categories = [
  {
    _id: '1',
    name: 'green',
  },
  {
    _id: '2',
    name: 'fruit',
  },
];

jest.mock('services/ceaiApi', () => ({
  getCategories: jest
    .fn()
    .mockImplementationOnce(() =>
      Promise.resolve({ success: true, message: 'Success', categories }),
    )
    .mockImplementationOnce(() =>
      Promise.resolve({ success: false, message: 'Not Found' }),
    ),
}));

describe('CategoryMenu', () => {
  beforeEach((getCategories as jest.Mock).mockClear);

  it('renders correctly', async () => {
    const { container, getAllByRole } = render(<CategoryMenu />);

    await waitFor(() => expect(getCategories).toBeCalledTimes(1));

    const categoriesList = getAllByRole('link');

    categoriesList.forEach((category, i) => {
      expect(category).toBeVisible();
      expect(category).toHaveTextContent(categories[i].name);
      expect(category).toHaveAttribute(
        'href',
        expect.stringContaining(categories[i].name),
      );
      expect(category).toHaveAttribute(
        'aria-label',
        expect.stringContaining(categories[i].name),
      );
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("doesn't render the list with no data", () => {
    const { queryAllByRole } = render(<CategoryMenu />);

    expect(queryAllByRole('link').length).toBe(0);
    expect(getCategories).toBeCalledTimes(1);
  });
});
