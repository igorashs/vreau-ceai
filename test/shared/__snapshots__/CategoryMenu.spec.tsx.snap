// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryMenu renders correctly 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 4px;
  box-shadow: none;
  padding: 0;
}

.c2:hover {
  background-color: var(--accent);
}

.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
  color: var(--accent-text-dark);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4:hover {
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
}

.c1 p {
  font-weight: 500;
}

.c1 button {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c3 {
  display: grid;
  -webkit-column-gap: var(--baseline);
  column-gap: var(--baseline);
  grid-template-columns: repeat(auto-fit,minmax(224px,1fr));
}

.c3 li {
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
}

.c0 h5 {
  margin-bottom: calc(var(--baseline) / 4);
}

<div
  class="c0"
>
  <h5>
    Categorii
  </h5>
  <div>
    <div
      class="c1"
    >
      <p>
        Ceai
      </p>
      <button
        class="c2"
      >
        <svg />
      </button>
    </div>
    <ul
      class="c3"
    >
      <li>
        <a
          aria-label="Categoria green"
          class="c4"
          href="/categories/green"
          rel=""
          target=""
        >
          green
        </a>
      </li>
      <li>
        <a
          aria-label="Categoria fruit"
          class="c4"
          href="/categories/fruit"
          rel=""
          target=""
        >
          fruit
        </a>
      </li>
    </ul>
  </div>
</div>
`;
